### PR TITLE
Custom partition support in topics

### PR DESCRIPTION
**Pull request by rutul25**

### CODE REVIEW

File: internal/topic/validation.go [hunk 1]
```diff
@@ -34,7 +34,12 @@ func GetValidatedModel(ctx context.Context, req *metrov1.Topic) (*Model, error)
 	m.Labels = req.GetLabels()
 	m.ExtractedProjectID = p
 	m.ExtractedTopicName = t
-	m.NumPartitions = DefaultNumPartitions
+	if req.NumOfPartitions == 0 {
+		m.NumPartitions = DefaultNumPartitions
+	} else {
+		m.NumPartitions = int(req.NumOfPartitions)
```
Comment (vnktram) [inline internal/topic/validation.go:40]: Please add a max limit here. We do not want an unsanitized input to be passed on.

Comment (vnktram) [reply internal/topic/validation.go:40]: Also please ensure that the topic entity in consul reflects the same number.

**Review by vnktram — CHANGES_REQUESTED** [review]

**Review by vnktram — APPROVED** [review]
LGTM

File: internal/tasks/schedulertask.go [hunk 2]
```diff
@@ -202,132 +223,172 @@ func (sm *SchedulerTask) Run(ctx context.Context) error {
...
+// 	 (Invalid due to changes in the subscription, node failures, topic updates, etc)
+//	  a. Remove bindings that do not conform to the new partition based approach.
+//    b. Remove nodebindings for deleted/invalid subscriptions.
+//    c. Remove nodebindings impacted by node failures
+// 2. Evaluate subscriptions and schedule any missing subscription/partition combos to nodes available.
+// 3. Topic changes are inherently covered since subscription validates against topic.
 func (sm *SchedulerTask) refreshNodeBindings(ctx context.Context) error {
+	err := sm.refreshCache(ctx)
+	if err != nil {
+		logger.Ctx(ctx).Errorw("schedulertask: Filed to refresh cache for topic/subscripiton/nodes", "error", err.Error())
```
Comment (sakshi-goyal-razorpay) [inline internal/tasks/schedulertask.go:279]: typo.. 
```suggestion
		logger.Ctx(ctx).Errorw("schedulertask: Failed to refresh cache for topic/subscription/nodes", "error", err.Error())
```

File: internal/tasks/schedulertask.go [hunk 3]
```diff
@@ -202,132 +223,172 @@ func (sm *SchedulerTask) Run(ctx context.Context) error {
...
+// refreshNodeBindings achieves the following in the order outline:
+// 1. Go through node bindings and remove invalid bindings
+// 	 (Invalid due to changes in the subscription, node failures, topic updates, etc)
+//	  a. Remove bindings that do not conform to the new partition based approach.
+//    b. Remove nodebindings for deleted/invalid subscriptions.
+//    c. Remove nodebindings impacted by node failures
+// 2. Evaluate subscriptions and schedule any missing subscription/partition combos to nodes available.
+// 3. Topic changes are inherently covered since subscription validates against topic.
 func (sm *SchedulerTask) refreshNodeBindings(ctx context.Context) error {
+	err := sm.refreshCache(ctx)
```
Comment (sakshi-goyal-razorpay) [inline internal/tasks/schedulertask.go:277]: why are we refreshing all three caches, if one is changed..?

File: internal/tasks/schedulertask.go [hunk 4]
```diff
@@ -202,132 +223,172 @@ func (sm *SchedulerTask) Run(ctx context.Context) error {
...
 	// fetch all current node bindings across all nodes
 	nodeBindings, err := sm.nodeBindingCore.List(ctx, nodebinding.Prefix)
+
+	if err != nil {
+		logger.Ctx(ctx).Errorw("error fetching new node binding list", "error", err)
+		return err
+	}
+	nodeBindingKeys, err := sm.nodeBindingCore.ListKeys(ctx, nodebinding.Prefix)
 	if err != nil {
 		logger.Ctx(ctx).Errorw("error fetching new node binding list", "error", err)
```
Comment (sakshi-goyal-razorpay) [inline internal/tasks/schedulertask.go:290]: need to change error message

File: internal/tasks/schedulertask.go (hunk 3 above)
Comment (vnktram) [reply internal/tasks/schedulertask.go:277]: The existing local cache might still be the one that holds non-partition-specific bindings due to the old code flow. We want to fetch the cache for all three entities and verify that the old bindings are removed and new bindings are added accordingly.

**Review by sakshi-goyal-razorpay — APPROVED** [review]

**Pull request merged**